### PR TITLE
fix referral set command

### LINKED ISSUE #9208: Refactor Team / referral set
Fix the Referral set API and state:
- implement the ReferralSet commands
- Keep track of the referral set in the referral engine
-

### PR DESCRIPTION
**Pull request by jeremyletang**
close #9208

### CODE REVIEW

File: commands/create_team.go [hunk 1]
```diff
@@ -2,16 +2,20 @@ package commands
...
+func CheckCreateReferralSet(cmd *commandspb.CreateReferralSet) error {
+	return checkCreateReferralSet(cmd).ErrorOrNil()
 }
 
-func checkCreateTeam(cmd *commandspb.CreateTeam) Errors {
+func checkCreateReferralSet(cmd *commandspb.CreateReferralSet) Errors {
 	errs := NewErrors()
 
 	if cmd == nil {
 		return errs.FinalAddForProperty("create_team", ErrIsRequired)
```
Comment (ValentinTrinque) [inline commands/create_team.go:13]:
```suggestion
		return errs.FinalAddForProperty("create_referral_set", ErrIsRequired)
```

File: commands/update_team.go [hunk 2]
```diff
@@ -2,20 +2,24 @@ package commands
...
+func CheckUpdateReferralSet(cmd *commandspb.UpdateReferralSet) error {
+	return checkUpdateReferralSet(cmd).ErrorOrNil()
 }
 
-func checkUpdateTeam(cmd *commandspb.UpdateTeam) Errors {
+func checkUpdateReferralSet(cmd *commandspb.UpdateReferralSet) Errors {
 	errs := NewErrors()
 
 	if cmd == nil {
 		return errs.FinalAddForProperty("update_team", ErrIsRequired)
```
Comment (ValentinTrinque) [inline commands/update_team.go:13]:
```suggestion
		return errs.FinalAddForProperty("update_referral_set", ErrIsRequired)
```

Comment (ValentinTrinque) [inline commands/update_team.go:13]:
```suggestion
		return errs.FinalAddForProperty("update_referral_set", ErrIsRequired)
```

File: core/referral/engine.go [hunk 3]
```diff
@@ -49,6 +66,77 @@ type Engine struct {
...
+
+	e.sets[deterministicID] = &types.ReferralSet{
+		ID:        deterministicID,
+		CreatedAt: now,
+		UpdatedAt: now,
+		Referrer: &types.Membership{
+			PartyID:  types.PartyID(party),
+			JoinedAt: now,
+			// TODO:
+			// not sure we need number of epoch
```
Comment (ValentinTrinque) [inline core/referral/engine.go:104]: Yes we do, but not as is, it should be a `StartAtEpoch`. There is an issue for that

**Review by ValentinTrinque — DISMISSED** [review]

File: core/referral/engine.go (hunk 3 above)
Comment (jeremyletang) [reply core/referral/engine.go:105]: alright, cool. let's fix it in that PR?

**Review by EVODelavega — DISMISSED** [review]
LGTM

File: protos/sources/vega/events/v1/events.proto [hunk 4]
```diff
@@ -727,6 +727,26 @@ message RefereeJoinedTeam {
   int64 joined_at = 3;
 }
 
+message ReferralSetCreated {
+  // The unique identifier of the created set.
+  string set_id = 1;
+  // The party that created the set.
+  string referrer = 2;
+  // Time in Unix nanoseconds when the set is created.
```
Comment (candida-d) [inline protos/sources/vega/events/v1/events.proto:735]:
```suggestion
  // Time in Unix nanoseconds when the set was created.
```

File: protos/sources/vega/events/v1/events.proto [hunk 5]
```diff
@@ -727,6 +727,26 @@ message RefereeJoinedTeam {
...
+  int64 created_at = 3;
+  // Time in Unix nanoseconds when the set is created.
```
Comment (candida-d) [inline protos/sources/vega/events/v1/events.proto:737]:
```suggestion
  // Time in Unix nanoseconds when the set was updated.
```

**Review by EVODelavega — DISMISSED** [review]

File: protos/sources/vega/snapshot/v1/snapshot.proto [hunk 6]
```diff
@@ -1090,6 +1092,18 @@ message NewReferralProgram {
...
+message ReferralSets {
+  repeated ReferralSet sets = 1;
+}
+
+message ReferralSet {
+  string id = 1;
+  int64 created_at = 2;
+  int64 updated_at = 3;
+  Membership referrer = 4;
+  repeated Membership referrees = 5;
```
Comment (candida-d) [inline protos/sources/vega/snapshot/v1/snapshot.proto:1104]: This should be "referees". I haven't checked for all instances.

File: protos/sources/vega/events/v1/events.proto [hunk 7]
```diff
@@ -1050,6 +1076,10 @@ message BusEvent {
     ReferralProgramUpdated referral_program_updated = 166;
     // Event notifying that a referral program has ended.
     ReferralProgramEnded referral_program_ended = 167;
+    // Event notifying a referral set has been created.
+    ReferralSetCreated referral_set_created = 168;
+    // Event notifying a referree has joined a referral set.
```
Comment (candida-d) [inline protos/sources/vega/events/v1/events.proto:1081]:
```suggestion
    // Event notifying a referee has joined a referral set.
```

File: protos/sources/vega/events/v1/events.proto [hunk 8]
```diff
@@ -727,6 +727,26 @@ message RefereeJoinedTeam {
   int64 joined_at = 3;
 }
 
+message ReferralSetCreated {
+  // The unique identifier of the created set.
```
Comment (candida-d) [inline protos/sources/vega/events/v1/events.proto:731]:
```suggestion
  // Unique ID of the created set.
```

File: protos/sources/vega/events/v1/events.proto [hunk 9]
```diff
@@ -727,6 +727,26 @@ message RefereeJoinedTeam {
...
+  string set_id = 1;
+  // The party that created the set.
```
Comment (candida-d) [inline protos/sources/vega/events/v1/events.proto:733]:
```suggestion
  // Party that created the set.
```

File: protos/sources/vega/events/v1/events.proto [hunk 10]
```diff
@@ -727,6 +727,26 @@ message RefereeJoinedTeam {
...
+  // Time in Unix nanoseconds when the set is created.
+  int64 created_at = 3;
+  // Time in Unix nanoseconds when the set is created.
+  int64 updated_at = 4;
+}
+
+message RefereeJoinedReferralSet {
+  // The unique identifier of the referral set the referee joined.
+  string set_id = 1;
+  // The party that joined the set.
```
Comment (candida-d) [inline protos/sources/vega/events/v1/events.proto:744]:
```suggestion
  // Unique ID of the referral set the referee joined.
  string set_id = 1;
  // Party that joined the set.
```

File: core/referral/engine.go [hunk 11]
```diff
@@ -167,29 +249,53 @@ func (e *Engine) loadNewReferralProgramFromSnapshot(program *vegapb.ReferralProg
...
+				JoinedAt:      time.Unix(0, set.Referrer.JoinedAt),
+				NumberOfEpoch: set.Referrer.NumberOfEpoch,
+			},
 		}
-	}
 
-	return nil
+		e.referrers[set.Referrer.PartyId] = set.Id
+
+		for _, r := range set.Referrees {
```
Comment (candida-d) [inline core/referral/engine.go:271]:
```suggestion
		for _, r := range set.Referees {
```

File: core/referral/snapshot.go [hunk 12]
```diff
@@ -83,15 +89,65 @@ func (e *SnapshottedEngine) serialise(k string) ([]byte, error) {
...
+			UpdatedAt: set.UpdatedAt.UnixNano(),
+			Referrer: &snapshotpb.Membership{
+				PartyId:       string(set.Referrer.PartyID),
+				JoinedAt:      set.Referrer.JoinedAt.UnixNano(),
+				NumberOfEpoch: set.Referrer.NumberOfEpoch,
+			},
+		}
+
+		for _, r := range set.Referees {
+			pset.Referrees = append(pset.Referrees,
```
Comment (candida-d) [inline core/referral/snapshot.go:119]:
```suggestion
			pset.Referees = append(pset.Referees,
```

File: protos/sources/vega/commands/v1/commands.proto [hunk 13]
```diff
@@ -304,31 +304,45 @@ message IssueSignatures {
...
+    optional string avatar_url = 12;
+    // Is this team closed for external other users
+    bool closed = 13;
+  }
 }
 
 // Request for updating a team's properties
-message UpdateTeam {
+message UpdateReferralSet {
   // Team ID of the team to update.
```
Comment (candida-d) [inline protos/sources/vega/commands/v1/commands.proto:331]:
```suggestion
  // ID of the referral set to update.
```

File: protos/sources/vega/commands/v1/commands.proto [hunk 14]
```diff
@@ -304,31 +304,45 @@ message IssueSignatures {
   string validator_node_id = 3; // the node for which to emit the signatures.
 }
 
-// Request for creating a referral team
+// Request for creating a referral set
 //
 // Creates a referral team. The team creator automatically becomes
 // the team leader, called a referrer. This cannot be changed.
 // A referrer cannot already be or become a liquidity provider,
 // nor can they be on an existing team as referrer or referee.
```
Comment (candida-d) [inline protos/sources/vega/commands/v1/commands.proto:312]:
```suggestion
// Creates a referral set. The creator automatically becomes
// the leader, called a referrer. This cannot be changed.
// A referrer cannot already be or become a liquidity provider,
// nor can they be part of an existing set as referrer or referee.
```

File: protos/sources/vega/commands/v1/commands.proto [hunk 15]
```diff
@@ -304,31 +304,45 @@ message IssueSignatures {
...
+    // Optional link to a team forum, discord, etc.
+    optional string team_url = 11;
+    // Optional link to an image to be used as the team avatar.
+    optional string avatar_url = 12;
+    // Is this team closed for external other users
+    bool closed = 13;
+  }
 }
 
 // Request for updating a team's properties
```
Comment (candida-d) [inline protos/sources/vega/commands/v1/commands.proto:329]:
```suggestion
// Request for updating a referral set's properties
```

File: protos/sources/vega/commands/v1/commands.proto [hunk 16]
```diff
@@ -304,31 +304,45 @@ message IssueSignatures {
...
+    // Optional link to a team forum, discord, etc.
+    optional string team_url = 11;
+    // Optional link to an image to be used as the team avatar.
+    optional string avatar_url = 12;
+    // Is this team closed for external other users
+    optional bool closed = 13;
+  }
 }
 
 // Request to join a team
```
Comment (candida-d) [inline protos/sources/vega/commands/v1/commands.proto:348]:
```suggestion
// Request to apply a referral code
```

File: protos/sources/vega/commands/v1/commands.proto [hunk 17]
```diff
@@ -337,7 +351,7 @@ message UpdateTeam {
 // be a liquidity provider, nor can they create a team or join multiple teams.
 // To switch teams, the referee can ask to join another team, and the switch will
 // be effective at the end of the epoch.
```
Comment (candida-d) [inline protos/sources/vega/commands/v1/commands.proto:353]:
```suggestion
// be a liquidity provider, nor can they create a referral set or join multiple sets.
// To switch, the referee can ask to join another set, and the switch will
// be effective at the end of the epoch.
```

Comment (candida-d) [inline protos/sources/vega/commands/v1/commands.proto:353]: The line above (353) also needs to change but I can't submit a comment on it

File: protos/sources/vega/commands/v1/transaction.proto [hunk 18]
```diff
@@ -73,11 +73,11 @@ message InputData {
     // Command to cancel stop orders.
     StopOrdersCancellation stop_orders_cancellation = 1017;
     // Command to create a team.
-    CreateTeam create_team = 1018;
+    CreateReferralSet create_referral_set = 1018;
     // Command to update a team.
```
Comment (candida-d) [inline protos/sources/vega/commands/v1/transaction.proto:77]:
```suggestion
    // Command to update a referral set.
```

File: protos/sources/vega/commands/v1/transaction.proto [hunk 19]
```diff
@@ -73,11 +73,11 @@ message InputData {
...
-    UpdateTeam update_team = 1019;
+    UpdateReferralSet update_referral_set = 1019;
     // Command to join a team.
```
Comment (candida-d) [inline protos/sources/vega/commands/v1/transaction.proto:79]:
```suggestion
    // Command to apply a referral code.
```

File: commands/apply_referral_code_test.go [hunk 20]
```diff
@@ -10,31 +10,31 @@ import (
 	"github.com/stretchr/testify/assert"
 )
 
-func TestJoinTeam(t *testing.T) {
+func TestApplyReferralCode(t *testing.T) {
 	t.Run("Joining team succeeds", testJoiningTeamSucceeds)
-	t.Run("Joining team with team ID fails", testJoinTeamWithoutTeamIDFails)
+	t.Run("Joining team with team ID fails", testApplyReferralCodeWithoutTeamIDFails)
```
Comment (candida-d) [inline commands/apply_referral_code_test.go:15]: How can this be true if there's no such thing as a team ID anymore?

File: protos/sources/vega/commands/v1/commands.proto [hunk 21]
```diff
@@ -337,7 +351,7 @@ message UpdateTeam {
...
-message JoinTeam {
+message ApplyReferralCode {
   // Team ID of the team to join.
```
Comment (candida-d) [inline protos/sources/vega/commands/v1/commands.proto:355]:
```suggestion
  // ID of the referral set to join.
```

**Review by candida-d — COMMENTED** [review]
Lots of mixed terminology in this PR's changes

File: commands/apply_referral_code_test.go (hunk 20 above)
Comment (jeremyletang) [reply commands/apply_referral_code_test.go:15]: it's just naming of a test, that's fine. teams still exists but in a different form

**Comment by jeremyletang** [comment]
@candida-d yes lots of them, we were planning to address the code comments in the next PR, but anyway. I've merged your changes and address the few referree -> referee changes.

**Comment by candida-d** [comment]
> @candida-d yes lots of them, we were planning to address the code comments in the next PR, but anyway. I've merged your changes and address the few referree -> referee changes.

Unfortunately I've not yet acquired mind-reading as a skill! So next a comment to that effect would help.

**Comment by jeremyletang** [comment]
> > @candida-d yes lots of them, we were planning to address the code comments in the next PR, but anyway. I've merged your changes and address the few referree -> referee changes.
> 
> Unfortunately I've not yet acquired mind-reading as a skill! So next a comment to that effect would help.

indeed, my bad.

File: protos/sources/vega/commands/v1/commands.proto [hunk 22]
```diff
@@ -304,40 +304,54 @@ message IssueSignatures {
...
 }
 
-// Request for creating a referral team
+// Request for creating a referral set
 //
-// Creates a referral team. The team creator automatically becomes
-// the team leader, called a referrer. This cannot be changed.
+// Creates a referral set. The creator automatically becomes
+// the leader, called a referrer. This cannot be changed.
 // A referrer cannot already be or become a liquidity provider,
```
Comment (cdummett) [inline protos/sources/vega/commands/v1/commands.proto:311]:
```suggestion
```

File: protos/sources/vega/commands/v1/commands.proto [hunk 23]
```diff
@@ -304,40 +304,54 @@ message IssueSignatures {
...
-// nor can they be on an existing team as referrer or referee.
-message CreateTeam {
-  // Optional team name to be added to the referral banner.
-  optional string name = 1;
-  // Optional link to a team forum, discord, etc.
-  optional string team_url = 2;
-  // Optional link to an image to be used as the team avatar.
-  optional string avatar_url = 3;
+// nor can they be part of an existing set as referrer or referee.
```
Comment (cdummett) [inline protos/sources/vega/commands/v1/commands.proto:312]:
```suggestion
// A referrer cannot be part of an existing set as referrer or referee.
```

File: protos/sources/vega/commands/v1/commands.proto [hunk 24]
```diff
@@ -304,40 +304,54 @@ message IssueSignatures {
...
 }
 
-// Request to join a team
+// Request to apply a referral code
 //
 // A party that joins a referral team is called a referee. A referee cannot
-// be a liquidity provider, nor can they create a team or join multiple teams.
-// To switch teams, the referee can ask to join another team, and the switch will
+// be a liquidity provider, nor can they create a referral set or join multiple sets.
+// To switch, the referee can ask to join another set, and the switch will
```
Comment (cdummett) [inline protos/sources/vega/commands/v1/commands.proto:352]:
```suggestion
// create a referral set or join multiple sets.
// To switch, the referee can ask to join another set, and the switch will
```

**Review by cdummett — COMMENTED** [review]
Docstrings still mention the removed restriction of LPs participating in the program.

**Pull request merged**